### PR TITLE
Enable executable and lib signatures in macOS

### PR DESCRIPTION
**Pull request by folmos-at-orange**
- Add to the CI conda workflow the necessary environment variables so the macOS packages are signed (only on tag). The necessary variables are all set to values from GitHub organization secrets.
- Add signature for the KNI library

### CODE REVIEW

File: .github/workflows/conda.yml [hunk 1]
```diff
@@ -53,19 +53,27 @@ jobs:
...
-          # The conda version cannot have '-' as a character so we eliminate it
           echo "KHIOPS_VERSION=$(echo $KHIOPS_RAW_VERSION | sed 's/-//')" >> "$GITHUB_ENV"
+
+          # On tag and macOS: Set the environment variables to sign the binaries
+          if [[ "${{ runner.os }}" == "macOS" && "${{ github.ref_type }}" == "tag" ]]
+          then
+            echo "KHIOPS_APPLE_CERTIFICATE_COMMON_NAME=${{ secrets.KHIOPS_APPLE_CERTIFICATE_COMMON_NAME }}" >> "$GITHUB_ENV"
+            echo "KHIOPS_APPLE_CERTIFICATE_BASE64=${{ secrets.KHIOPS_APPLE_CERTIFICATE_BASE64 }}" >> "$GITHUB_ENV"
+            echo "KHIOPS_APPLE_CERTIFICATE_PASSWORD=${{ secrets.KHIOPS_APPLE_CERTIFICATE_PASSWORD }}" >> "$GITHUB_ENV"
+            echo "KHIOPS_APPLE_TMP_KEYCHAIN_PASSWORD=${{ secrets.KHIOPS_APPLE_TMP_KEYCHAIN_PASSWORD }}" >> "$GITHUB_ENV"
```
Comment (popescu-v) [inline .github/workflows/conda.yml:75]: I would have done this directly in an `env:` section of the steps where these secrets are needed:
```
env:
  KHIOPS_APPLE_CERTIFICATE_COMMON_NAME: ${{ secrets.KHIOPS_APPLE_CERTIFICATE_COMMON_NAME }}
```

etc, as these secrets are not _computed_ in this step, they come from GitHub settings; so, we can add them to the `env:` as we whish; see also e.g. https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions#example-using-bash

File: packaging/conda/build.sh [hunk 2]
```diff
@@ -15,7 +15,8 @@ fi
 cmake --fresh --preset $CMAKE_PRESET -DBUILD_JARS=OFF -DTESTING=OFF
 
 # Build MODL and MODL_Coclustering
-cmake --build --preset $CMAKE_PRESET --parallel --target MODL MODL_Coclustering KhiopsNativeInterface KNITransfer 
+cmake --build --preset $CMAKE_PRESET --parallel \
+  --target MODL MODL_Coclustering KhiopsNativeInterface KNITransfer
```
Comment (popescu-v) [inline packaging/conda/build.sh:19]: This change does not seem related to the current PR; was it required by the pre-commit hooks?

File: .github/workflows/conda.yml (hunk 1 above)
Comment (folmos-at-orange) [reply .github/workflows/conda.yml:75]: I want to add them just when they are necessary (macOS + tag). Before I used an `env:` in the proper step, but the necessary `${{ ...}}` statements were very cryptic. This is clearer and easier to debug.

**EDIT*** In addition If the variable `KHIOPS_CERTIFICATE_COMMON_NAME` is defined it triggers the signature. There is no way to "not define" a variable in an `env:` statement.

File: packaging/conda/build.sh (hunk 2 above)
Comment (folmos-at-orange) [reply packaging/conda/build.sh:19]: No, just a small editorial change because the line was too long.

File: packaging/conda/build.sh [hunk 3]
```diff
@@ -101,11 +103,16 @@ then
   # identity and makes the build fail!
   CODESIGN="/usr/bin/codesign"
 
-  # Sign the MODL executable and check
+  # Obtain the path of the real KNI (not the symlinks)
+  KNI_PATH=$(file --mime --no-dereference $PREFIX/lib/libKhiops* | grep x-mach-binary | cut -d: -f1)
```
Comment (popescu-v) [inline packaging/conda/build.sh]: I'm not sure I understand this well; some DLLs give us `x-sharedlib`. What do we want to achieve here?

Comment (folmos-at-orange) [reply packaging/conda/build.sh]: In macOS this is the output of the command is
```bash
(kh-dev) folmos@MC-JWMHQQ2JYL:~/work/khiops/khiops (add-certs-to-conda)
(11:59:36)> file -I --no-dereference  ~/Downloads/khiops-conda-osx-arm64-4/osx-arm64/kni-10.3.0a.2-h2833df3_0/lib/libKhiopsNativeInterface.*
/Users/folmos/Downloads/khiops-conda-osx-arm64-4/osx-arm64/kni-10.3.0a.2-h2833df3_0/lib/libKhiopsNativeInterface.10.2.2-b.2.dylib: application/x-mach-binary; charset=binary
/Users/folmos/Downloads/khiops-conda-osx-arm64-4/osx-arm64/kni-10.3.0a.2-h2833df3_0/lib/libKhiopsNativeInterface.10.dylib:         inode/symlink; charset=binary
/Users/folmos/Downloads/khiops-conda-osx-arm64-4/osx-arm64/kni-10.3.0a.2-h2833df3_0/lib/libKhiopsNativeInterface.dylib:            inode/symlink; charset=binary
```
So the mime type of dynamic libraries in macOS is `application/x-mach-binary`

File: .github/workflows/conda.yml (hunk 1 above)
Comment (popescu-v) [reply .github/workflows/conda.yml:75]: Yeah, I see. Moreover, there is also no way to unset the env vars from `GITHUB_ENV` either...

File: packaging/conda/build.sh (hunk 2 above)
Comment (popescu-v) [reply packaging/conda/build.sh:19]: Ok.

File: packaging/conda/build.sh (hunk 3 above)
Comment (popescu-v) [reply packaging/conda/build.sh]: Perhaps `readlink -f` might help achieve the goal here?

Comment (folmos-at-orange) [reply packaging/conda/build.sh]: I could change to `KNI_PATH=$(readlink -f $PREFIX/lib/libKhiops* | uniq)`

Comment (popescu-v) [reply packaging/conda/build.sh]: Yeah, it seems more readable and straightforward IMHO.

Comment (popescu-v) [reply packaging/conda/build.sh]: Perhaps add ` | head -n 1` to ensure that structurally there is a unique entry?

Comment (folmos-at-orange) [reply packaging/conda/build.sh]: Ok & Done.

Comment (popescu-v) [reply packaging/conda/build.sh]: I don't see  the ` | head -n 1`...

Comment (folmos-at-orange) [reply packaging/conda/build.sh]: It should be one, if not, there is a problem and better not to hide it.

Comment (popescu-v) [reply packaging/conda/build.sh]: Oh, ok then the build would eventually fail because of this, if the KNI path is not unique.

**Review by popescu-v — COMMENTED** [review]
Rebase current branch on `dev-v10`. Then cherry-pick on `dev` as well.

**Review by popescu-v — APPROVED** [review]
LGTM
